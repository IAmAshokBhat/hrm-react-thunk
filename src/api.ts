import {
  ILoginData,
  INewAttendance,
  INewLeave,
  IRegisterAPI
} from './constants';
import { fetchWithToken, getToken } from './utils';

const headers = {
  'X-JWT-Token': getToken()
};

export const getUserListAPI = () => fetchWithToken('user/list');

export const getUserDetailsAPI = (id: number) =>
  fetchWithToken(`api/users/${id}`);

export const getAttendanceDetailsAPI = (id: number, from: string, to: string) =>
  fetchWithToken(`api/attendance/${id}?from=${from}&to=${to}`);

export const getLeaveBalanceAPI = (id: number) =>
  fetchWithToken(`api/leaves/${id}`);

export const getEventDetailsAPI = (id: number, from: string, to: string) =>
  fetchWithToken(`api/events/${id}?from=${from}&to=${to}`);

export const getAllLeavesAPI = (id: number) =>
  fetchWithToken(`api/all-leaves/${id}`);

export const applyLeaveAPI = (newLeave: INewLeave) =>
  fetchWithToken(`api/apply-leave`, {
    method: 'POST',
    body: JSON.stringify(newLeave)
  });

export const addAttendanceAPI = (newAttendance: INewAttendance) =>
  fetchWithToken(`api/attendance`, {
    method: 'POST',
    body: JSON.stringify(newAttendance)
  });

export const loginAPI = (userInfo: ILoginData) =>
  fetchWithToken(`login`, {
    method: 'POST',
    body: JSON.stringify(userInfo)
  });

export const getDesignationListAPI = () =>
  fetchWithToken(`designation`, { headers: headers });

export const registerAPI = (newUser: IRegisterAPI) =>
  fetchWithToken(`user`, {
    method: 'POST',
    body: JSON.stringify(newUser)
  });
