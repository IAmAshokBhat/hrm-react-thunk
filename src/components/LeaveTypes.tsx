import Paper from '@mui/material/Paper';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableCell';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { ILeaveBalance } from '../constants';
import { fetchLeaveBalanceAction } from '../redux/actions';
import { selectLeaveBalance } from '../redux/selectors';
import { getUserId } from '../utils';

export const LeaveTypes = () => {
  const dispatch = useDispatch();
  const leaveBalance: ILeaveBalance[] = useSelector(selectLeaveBalance);
  const loggedInUser = getUserId();

  useEffect(() => {
    dispatch(fetchLeaveBalanceAction(loggedInUser));
  }, [dispatch, loggedInUser]);

  return (
    <TableContainer component={Paper}>
      <Table sx={{ minWidth: 650 }} aria-label="simple table">
        <TableHead>
          <TableRow>
            <TableCell>Leave Type</TableCell>
            <TableCell align="right">Balance</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {leaveBalance.map(({ leaveType, balance }) => (
            <TableRow
              key={leaveType}
              sx={{ '&:last-child td, &:last-child th': { border: 0 } }}
            >
              <TableCell component="th" scope="row">
                {leaveType}
              </TableCell>
              <TableCell component="th" scope="row">
                {balance}
              </TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  );
};
