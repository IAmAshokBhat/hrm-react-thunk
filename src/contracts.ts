import { AnyAction, Dispatch } from 'redux';
import { TAny, TAnyObject } from './constants';

export type TFPromise = (params: TAny) => Promise<TAnyObject>;

export type TFPromiseAPI = (params?: TAny | TAnyObject) => Promise<TAnyObject>;

export type TFAction = (dispatch: Dispatch<AnyAction>) => Promise<void>;

/*
()
(1,2,3)
({key:value})
**
*/
