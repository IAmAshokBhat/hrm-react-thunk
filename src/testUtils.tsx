import { render, RenderOptions, waitFor } from '@testing-library/react';
import { SnackbarProvider } from 'notistack';
import { ReactElement } from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import { applyMiddleware, createStore, Store } from 'redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { IDefaultRoot, TAny, TAnyObject } from './constants';
import { TFAction } from './contracts';
import { reducers } from './redux/reducers';
import {
  defaultArrayState,
  defaultNullState,
  defaultState
} from './redux/reducers/__tests__/fixtures';

export const middlewares = [thunk];

export const createStoreWithMiddlewares = (
  initialState: TAnyObject = {}
): Store => {
  return createStore(reducers, initialState, applyMiddleware(...middlewares));
};
interface IRenderWithProviderOptions {
  initialStoreState?: TAnyObject;
  renderFunction?: TAny;
  options?: RenderOptions;
  route?: string;
  path?: string;
}

export const renderWithProvider = (
  ui: ReactElement,
  renderOptions?: IRenderWithProviderOptions
) => {
  const {
    initialStoreState = {},
    renderFunction = render,
    options = {},
    route = '',
    path = ''
  } = renderOptions || {};

  const appStore = createStoreWithMiddlewares(initialStoreState);

  let wrappedUi = (
    <SnackbarProvider
      anchorOrigin={{
        vertical: 'top',
        horizontal: 'right'
      }}
    >
      <Provider store={appStore}>{ui}</Provider>
    </SnackbarProvider>
  );

  if (route) {
    window.history.pushState({}, '', route);

    wrappedUi = (
      <Provider store={appStore}>
        <BrowserRouter>
          <Routes>
            <Route path={path} element={ui} />
          </Routes>
        </BrowserRouter>
      </Provider>
    );
  }

  return renderFunction(wrappedUi, options);
};

type TFApiCall = (...params: TAny) => Promise<Response>;

export interface IResponseMock {
  headers?: object;
  ok?: boolean;
  status?: number;
  statusText?: string;
  type?: ResponseType;
  url?: string;
  body?: TAny | null;
  bodyUsed?: boolean;
  blob?: () => Promise<TAny>;
}

const dummyPromise = () => new Promise((resolve) => resolve('dummy'));

export class ResponseMock {
  headers: object;

  ok: boolean;

  status: number;

  statusText: string;

  type: ResponseType;

  url: string;

  body: ReadableStreamDefaultReadResult<Uint8Array> | null;

  bodyUsed: boolean;

  blob: () => Promise<TAny>;

  constructor(props: IResponseMock = {}) {
    this.headers = props.headers || {};
    this.status = props.status || 200;
    this.ok = props.ok || this.status === 200;
    this.statusText = props.statusText || 'OK';
    this.type = props.type || 'basic';
    this.url = props.url || 'http://localhost';
    this.body = props.body;
    this.bodyUsed = props.bodyUsed || this.body !== null;
    this.blob = props.blob || dummyPromise;
  }

  json: () => Promise<TAny> = () => {
    const body = this.body || '""';

    try {
      return Promise.resolve(JSON.parse(body as string));
    } catch {
      return Promise.resolve(body);
    }
  };

  text: () => Promise<string> = () => Promise.resolve(this.statusText);
}

export const mockAPICall = (func: TFApiCall, response: IResponseMock) =>
  (func as jest.Mock).mockResolvedValue(new ResponseMock(response));

export const generateActionTests = async (
  actionName: string,
  action: TFAction,
  api: TAny,
  actionPayload: TAny = []
) => {
  const response = new ResponseMock({ ok: true, body: { data: [] } });

  const mockStore = configureStore(middlewares);

  (api as jest.Mock).mockReturnValue(Promise.resolve(response));
  const store = mockStore({});

  //TODO: Fix TS error
  // @ts-ignore
  await store.dispatch(action);

  await waitFor(() => {
    expect(store.getActions()).toEqual([
      {
        type: `${actionName}_LOADING`
      },
      {
        type: `${actionName}_SUCCESS`,
        //TODO: Accept action params from function params
        payload: actionPayload
      }
    ]);
  });
};

export const generateReducerTests = (
  reducerName: string,
  reducer: TAny,
  selector: TAny,
  actionName: string,
  successValue?: TAny,
  defaultValue?: TAny,
  expectedValue?: TAny,
  reducerLabel?: string
) => {
  describe(`for ${reducerName}`, () => {
    const reducerLbl = reducerLabel
      ? reducerLabel
      : reducerName.replace('Reducer', '');
    it(`${reducerName} reducer loading`, () => {
      reducer(undefined, {
        type: `${actionName}_LOADING`,
        payload: null
      });
      // TODO: Check why its failing
      // expect(selector(defaultState).loading).toEqual(true);
      expect(selector(defaultState).value).toEqual(
        defaultValue === null ? defaultValue : []
      );
      expect(selector(defaultState).error).toEqual(null);
    });

    it(`${reducerName} reducer success`, () => {
      const successState: IDefaultRoot = {
        ...defaultState,
        [reducerLbl]: {
          ...defaultNullState,
          value: expectedValue ? expectedValue : successValue
        }
      };
      reducer(defaultArrayState, {
        type: `${actionName}_SUCCESS`,
        payload: successValue
      });

      expect(selector(successState).error).toEqual(null);
      expect(selector(successState).loading).toEqual(false);
      expect(selector(successState).value).toEqual(
        expectedValue ? expectedValue : successValue
      );
    });

    it(`${reducerName} reducer error`, () => {
      const error = { message: '', status: 1, ok: false };
      const failedState: IDefaultRoot = {
        ...defaultState,
        [reducerLbl]: {
          ...defaultNullState,
          value: [],
          error
        }
      };

      reducer(undefined, {
        type: `${actionName}_FAILURE`,
        payload: error
      });

      expect(selector(failedState).loading).toEqual(false);
      expect(selector(failedState).value).toEqual([]);
      expect(selector(failedState).error).toEqual(error);
    });
  });
};
