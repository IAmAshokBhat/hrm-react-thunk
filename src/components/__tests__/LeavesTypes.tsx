import { getLeaveBalanceAPI } from '../../api';
import { TAnyObject } from '../../contracts';
import { mockAPICall, renderWithProvider } from '../../testUtils';
import { LeaveTypes } from '../LeaveTypes';
import { defaultState, leaveBalanceAPIResponse } from './fixtures';

jest.mock('../../api', () => ({
  getAllLeavesAPI: jest.fn(),
  getLeaveBalanceAPI: jest.fn()
}));

const LeaveBalanceResponse = {
  ok: true,
  body: leaveBalanceAPIResponse
};

describe('Leaves Page', () => {
  const generateNode = (initialStoreState: TAnyObject = defaultState) =>
    renderWithProvider(<LeaveTypes />, {
      initialStoreState
    });

  beforeEach(() => {
    mockAPICall(getLeaveBalanceAPI, LeaveBalanceResponse);
  });

  it('should render and match snapshot', () => {
    const { container } = generateNode();

    expect(container).toMatchSnapshot();
  });
});
