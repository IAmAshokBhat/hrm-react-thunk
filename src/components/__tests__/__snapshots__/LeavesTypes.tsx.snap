// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaves Page should render and match snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableCell-root MuiTableCell-sizeMedium css-sow7bl-MuiPaper-root-MuiTableCell-root"
  >
    <table
      aria-label="simple table"
      class="MuiTable-root css-quj9j6-MuiTable-root"
    >
      <thead
        class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          >
            Leave Type
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
            scope="col"
          >
            Balance
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
      />
    </table>
  </div>
</div>
`;
